### PR TITLE
NFCT cleanup test refactor

### LINKED ISSUE #856: Implement outstanding todos in NFCT

### LINKED ISSUE #858: Replace isValid with a modifier in PricelessPositionManager NFCT
right now each position stores an `isValid` variable. this can be replaced with a modifier that checks if the position's collateral is `>0`. This modifier should be applied to all methods except create and deposit.

### LINKED ISSUE #861: Consider removing require around transfer from methods
Right now all transfers are wrapped with a `require`. For example:
```
require(collateralCurrency.transfer(msg.sender, payToDisputer.rawValue), "failed to transfer reward for a successful dispute to disputer");
```

This is not strictly needed and removing the `require` will reduce code footprint.

### LINKED ISSUE #866: Comment NFCT with correct commenting guide

### LINKED ISSUE #874: Wrap constructor parameters in a struct to enable more variables to be passed
Right now we are limited in the number of parameters that can be passed via a constructor. This can be overcome by passing a struct within the constructor which contains all relevant variables.

### LINKED ISSUE #875: Add relevant events to the Expiring Multiparty contract

### LINKED ISSUE #887: Refactor `liquidatable.js` tests to use `toBN` over `bignumber.js`
Right now `liquidatable.js` uses 
```
const BN = require("bignumber.js");
```

for big numbers. this should be consistent with the other tests and use:

```
const { toBN } = web3.utils;
```

### PR DESCRIPTION
**Pull request by chrismaree**
This PR addresses a number of TODO's through the NFCT including refactoring, events, styling, contract ordering and commenting. The following issues are closed in this PR:
close #875 
close #856 
close #858 
close #874 
close #861 
close #866
close #887

### CODE REVIEW

File: core/contracts/financial_templates/Liquidatable.sol [hunk 1]
```diff
@@ -91,9 +61,31 @@ contract Liquidatable is PricelessPositionManager {
...
+        uint disputeBondAmount
+    );
+    event DisputeSettled(
+        address caller,
+        address sponsor,
+        address liquidator,
+        address disputer,
+        uint disputeId,
+        bool DisputeSucceeded
+    );
```
Comment (ptare) [inline core/contracts/financial_templates/Liquidatable.sol]: Can/should some of these params be indexed?

File: core/test/financial_templates/ExpiringMultiParty.js [hunk 2]
```diff
@@ -1,29 +1,36 @@
...
+      syntheticName: "Test UMA Token",
+      syntheticSymbol: "UMATEST",
+      liquidationLiveness: "1000",
+      collateralRequirement: { rawValue: toWei("1.5") },
+      disputeBondPct: { rawValue: toWei("0.1") },
+      sponsorDisputeRewardPct: { rawValue: toWei("0.1") },
+      disputerDisputeRewardPct: { rawValue: toWei("0.1") }
+    };
+    await ExpiringMultiParty.new(constructorParams);
+    assert(true);
```
Comment (ptare) [inline core/test/financial_templates/ExpiringMultiParty.js]: Do we need this assert here? I.e., if the deployment fails, would something already fail or no?

File: core/contracts/financial_templates/Liquidatable.sol (hunk 1 above)
Comment (chrismaree) [reply core/contracts/financial_templates/Liquidatable.sol]: We can for sure index some. Will refactor and add logical indices.

File: core/test/financial_templates/ExpiringMultiParty.js (hunk 2 above)
Comment (chrismaree) [reply core/test/financial_templates/ExpiringMultiParty.js]: I think you right; it's not required.

File: ci/run_slither.sh [hunk 3]
```diff
@@ -11,7 +11,7 @@ run_slither() {
     truffle compile 
 
     cd $PROTOCOL_DIR
-    slither --exclude=naming-convention,solc-version,pragma,external-function,reentrancy-benign,reentrancy-no-eth,arbitrary-send,locked-ether,reentrancy-eth,uninitialized-state-variables,incorrect-equality,reentrancy-events,assembly $1
+    slither --exclude=naming-convention,solc-version,pragma,external-function,reentrancy-benign,reentrancy-no-eth,arbitrary-send,locked-ether,reentrancy-eth,uninitialized-state-variables,incorrect-equality,reentrancy-events,assembly,shadowing-local $1
```
Comment (ptare) [inline ci/run_slither.sh]: I've been meaning to try Slither's triage mode [1] at some point, but haven't gotten around to it yet.

Definitely not part of this PR, but maybe we should add a task for triage Slither to the backlog?

1. https://github.com/crytic/slither/wiki/Usage#triage-mode

File: core/contracts/financial_templates/ExpiringMultiParty.sol [hunk 4]
```diff
@@ -6,33 +6,5 @@ import "../FixedPoint.sol";
 
 contract ExpiringMultiParty is Liquidatable {
     using FixedPoint for FixedPoint.Unsigned;
```
Comment (ptare) [inline core/contracts/financial_templates/ExpiringMultiParty.sol]: Do you need this `using` line?

File: core/test/financial_templates/ExpiringMultiParty.js [hunk 5]
```diff
@@ -1,29 +1,35 @@
...
 const truffleContract = require("@truffle/contract");
 const ERC20Mintable = truffleContract(ERC20MintableData);
 ERC20Mintable.setProvider(web3.currentProvider);
 
 contract("ExpiringMultiParty", function(accounts) {
-  it("Empty", async function() {
-    finder = await Finder.new();
+  it("Can deploy", async function() {
+    const collateralToken = await ERC20Mintable.new({ from: accounts[0] });
+    const finder = await Finder.new();
```
Comment (ptare) [inline core/test/financial_templates/ExpiringMultiParty.js]: I know you didn't change this line, but could we make this `await Finder.deployed()` instead? Just wondering actually.

File: core/test/TokenizedDerivative.js [hunk 6]
```diff
@@ -42,7 +42,7 @@ contract("TokenizedDerivative", function(accounts) {
   const apDelegate = accounts[3];
 
   const name = "1x Bitcoin-Ether";
-  const symbol = "BTCETH";
+  const symbol = "UMAETHUSD";
```
Comment (ptare) [inline core/test/TokenizedDerivative.js]: Why this unrelated change?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 7]
```diff
@@ -11,149 +11,173 @@ import "../Finder.sol";
...
+        string memory _syntheticSymbol
     ) public Testable(_isTest) {
         expirationTimestamp = _expirationTimestamp;
         withdrawalLiveness = _withdrawalLiveness;
         collateralCurrency = IERC20(_collateralAddress);
-        // TODO: add parameters to register the synthetic token's name and symbol.
-        // TODO: add the collateral requirement. This is needed at settlement and at dispute resolution.
-        // TODO: validate the input of the finder/price identifier inputs.
-        Token mintableToken = new Token();
+        Token mintableToken = new Token(_syntheticName, _syntheticSymbol, 18);
```
Comment (ptare) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: I see that this works, but it's very surprising to me that you don't need to inherit from ExpandedIERC20 in order to save mintableToken as an ExpandedIERC20 :/

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 8]
```diff
@@ -11,149 +11,173 @@ import "../Finder.sol";
 import "../OracleInterface.sol";
 /**
  * @title Financial contract with priceless position management.
- * @dev Handles positions for multiple sponsors in an optimistic (i.e., priceless) way without relying on a price feed.
+ * @author umaproject.org
```
Comment (ptare) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: We haven't previously used these author lines. Just wondering what others think of starting to use them everywhere.

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 9]
```diff
@@ -11,149 +11,173 @@ import "../Finder.sol";
...
+            "Deposit collateral currency transfer failed"
+        );
+
+        emit Deposit(msg.sender, collateralAmount.rawValue);
     }
 
     /**
      * @notice Transfers `collateralAmount` of `collateralCurrency` from the calling sponsor's position to the caller.
-     * Reverts if the withdrawal puts this position's collateralization ratio below the global collateralization ratio.
+     * @dev * Reverts if the withdrawal puts this position's collateralization ratio below the global collateralization ratio.
```
Comment (ptare) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: Extra `*`

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 10]
```diff
@@ -294,9 +376,22 @@ contract PricelessPositionManager is Testable {
...
+        onlyCollateralizedPosition(sponsor)
+        returns (PositionData storage)
+    {
+        return positions[sponsor];
+    }
+
+    function _getIdentifierWhitelistAddress() internal view returns (address) {
+        bytes32 identifierWhitelistInterface = "IdentifierWhitelist";
+        return finder.getImplementationAddress(identifierWhitelistInterface);
+    }
```
Comment (ptare) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: Surprised the linter isn't complaining about needing an empty line here? Or is this just the github diff being weird?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 11]
```diff
@@ -11,149 +11,173 @@ import "../Finder.sol";
...
+        _;
+    }
+
+    modifier onlyPostExpiration() {
+        require(getCurrentTime() >= expirationTimestamp, "Cannot operate on a position before its expiry time");
+        _;
+    }
+
+    modifier onlyCollateralizedPosition(address sponsor) {
+        require(positions[sponsor].collateral.rawValue > 0, "Position has no collateral and so is invalid");
```
Comment (ptare) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: More for my own understanding, do we pay multiple storage read costs if we read the same fields in this modifier and then again in each method?

File: core/test/financial_templates/PricelessPositionManager.js [hunk 12]
```diff
@@ -21,23 +25,28 @@ contract("PricelessPositionManager", function(accounts) {
...
+  // Contracts
   let collateral;
   let pricelessPositionManager;
   let tokenCurrency;
-  let priceTrackingIdentifier;
   let identifierWhitelist;
   let mockOracle;
   let finder;
 
+  //Initial constant values
```
Comment (ptare) [inline core/test/financial_templates/PricelessPositionManager.js]: ultra-nit: space after `//`

File: core/test/financial_templates/PricelessPositionManager.js [hunk 13]
```diff
@@ -70,27 +78,40 @@ contract("PricelessPositionManager", function(accounts) {
...
   });
 
-  // TODO: refactor these tests to use the `describe` keyword to break up each test into more modular and granular components.
+  it("Correct deployment and variable assignment", async function() {
+    // PricelessPosition variables
+    assert.equal(await pricelessPositionManager.expirationTimestamp(), expirationTimestamp);
+    assert.equal(await pricelessPositionManager.withdrawalLiveness(), withdrawalLiveness);
+    assert.equal(await pricelessPositionManager.collateralCurrency(), collateral.address);
+    assert.equal(await pricelessPositionManager.finder(), finder.address);
+    assert.equal(await pricelessPositionManager.priceIdentifer(), priceTrackingIdentifier.padEnd(66, 0));
```
Comment (ptare) [inline core/test/financial_templates/PricelessPositionManager.js]: Why not just check the string() versions? I.e., call `hexToUtf8`?

File: core/test/financial_templates/PricelessPositionManager.js [hunk 14]
```diff
@@ -147,7 +176,16 @@ contract("PricelessPositionManager", function(accounts) {
     assert(await didContractThrow(pricelessPositionManager.redeem({ rawValue: redeemTokens }, { from: sponsor })));
     await tokenCurrency.approve(pricelessPositionManager.address, redeemTokens, { from: sponsor });
     sponsorInitialBalance = await collateral.balanceOf(sponsor);
-    await pricelessPositionManager.redeem({ rawValue: redeemTokens }, { from: sponsor });
+    let redemptionResult = await pricelessPositionManager.redeem({ rawValue: redeemTokens }, { from: sponsor });
```
Comment (ptare) [inline core/test/financial_templates/PricelessPositionManager.js]: Are these variables reassigned?

File: core/contracts/financial_templates/Liquidatable.sol [hunk 15]
```diff
@@ -129,58 +121,62 @@ contract Liquidatable is PricelessPositionManager {
...
+            params.syntheticName,
+            params.syntheticSymbol
         )
     {
+        require(params.collateralRequirement.isGreaterThan(1), "The collateral requirement must be at minimum 100%");
         require(
-            _sponsorDisputeRewardPct.add(_disputerDisputeRewardPct).isLessThan(1),
-            "Sponsor and Disputer dispute rewards shouldn't sum to 100% or more"
+            params.sponsorDisputeRewardPct.add(params.disputerDisputeRewardPct).isLessThan(1),
+            "Dispute rewards should sum to > 100%"
```
Comment (ptare) [inline core/contracts/financial_templates/Liquidatable.sol]: Did you mean `<`?

File: core/contracts/financial_templates/Liquidatable.sol [hunk 16]
```diff
@@ -237,15 +244,19 @@ contract Liquidatable is PricelessPositionManager {
...
         // Enqueue a request with the DVM.
         _requestOraclePrice(disputedLiquidation.disputeTime);
+
+        emit LiquidationDisputed(sponsor, disputedLiquidation.liquidator, msg.sender, id, disputeBondAmount.rawValue);
     }
 
     /**
-     * Anyone can call this method to settle a pending dispute. This
-     * is only possible after the DVM has resolved a price. Callers should
+     * @notice After a liquidation has been disputed, it can be settled by any caller enabling withdrawl to occure.
```
Comment (ptare) [inline core/contracts/financial_templates/Liquidatable.sol]: typo in `occure`

File: core/contracts/financial_templates/Liquidatable.sol [hunk 17]
```diff
@@ -8,71 +8,41 @@ import "../FixedPoint.sol";
...
-        FixedPoint.Unsigned settlementPrice;
-    }
 
-    /**
-    * Contract-wide variables, consistent across all liquidations for synthetic tokens
-    * of this template
-    */
+        // Following variables set upon a dispute request
+        address disputer; // Person who is disputing a liquidation
+        uint disputeTime; // Time when dispute is initiated, needed to get price from Oracle
```
Comment (ptare) [inline core/contracts/financial_templates/Liquidatable.sol]: I know you didn't change this, but aren't Oracle requests supposed to be made at liquidationTime not disputeTime?

File: core/contracts/financial_templates/ExpiringMultiParty.sol (hunk 4 above)
Comment (chrismaree) [reply core/contracts/financial_templates/ExpiringMultiParty.sol]: I dont think it's needed in the `ExpiringMultiParty.sol`, can remove it.

File: core/test/financial_templates/ExpiringMultiParty.js (hunk 5 above)
Comment (chrismaree) [reply core/test/financial_templates/ExpiringMultiParty.js]: So this touches on a larger question on how we want to orchestrate tests. If we use `.deployed()` it will pull from the truffle migration version, as done in other tests within the stack. However, this makes the tests dependent on the migrations which right now the rest of these new tests are not.

Ultimately we need to decide if we want to deploy these helper like contracts in migrations then import like this with `deployed()` or re-create the contract setup in every test. There are tradeoffs between speed and reproducibility here.

File: core/test/TokenizedDerivative.js (hunk 6 above)
Comment (chrismaree) [reply core/test/TokenizedDerivative.js]: I updated the comments in how the tests flow and changed some of their magic numbers away from a Synthetic `BTCETH` pair as this was too complex to reason about in the maths.

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 7 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: Ye, for sure. In fact I think we can remove `ExpandedIERC20` from all places in this contract and just use `Token` everywhere for the synthetic token. I have made this change and it works fine and I think it is more logical.

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 8 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: ok, cool. thanks for the heads up. I removed it for consistency. can add it to all contracts later if we want.

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 10 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: actually it was missing a blank line. strange that the linter didnt catch it.

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 11 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: I actually have no idea. @mrice32 ? I will look into this as well

File: core/contracts/financial_templates/Liquidatable.sol (hunk 17 above)
Comment (mrice32) [reply core/contracts/financial_templates/Liquidatable.sol]: +1

File: core/test/financial_templates/PricelessPositionManager.js (hunk 14 above)
Comment (chrismaree) [reply core/test/financial_templates/PricelessPositionManager.js]: no, you right. will turn into a `const`

File: core/test/financial_templates/ExpiringMultiParty.js (hunk 5 above)
Comment (mrice32) [reply core/test/financial_templates/ExpiringMultiParty.js]: In #869, I move all of these over to `Finder.deployed()` because that PR requires the Finder to link the contract to the Store, etc.

File: core/contracts/financial_templates/Liquidatable.sol (hunk 17 above)
Comment (chrismaree) [reply core/contracts/financial_templates/Liquidatable.sol]: This is a really good point actually. This is wrong here and in the `createLiquidation` function. I will postpone this for a later PR and have created an issue #895

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 11 above)
Comment (mrice32) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: Yes, I think we do. I can't find any good references for this, but as far as I'm aware each SLOAD opcode gets charged 800 gas. The SSTORE opcode is the only one that gets special net-metering treatment. The only thing that might save us is if the solidity optimizer is smart and caches the value in memory (I highly doubt it's that smart, though).

Comment (ptare) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: Interesting, thanks for researching. We don't care about minimizing gas costs at this stage, so nothing to change here. I think the modifier is a win for readability.

File: core/contracts/FixedPoint.sol [hunk 18]
```diff
@@ -21,6 +21,11 @@ library FixedPoint {
         return Unsigned(a.mul(FP_SCALING_FACTOR));
     }
 
+    /** @dev Whether `a` is equal to `b`. */
+    function isEqual(Unsigned memory a, Unsigned memory b) internal pure returns (bool) {
```
Comment (mrice32) [inline core/contracts/FixedPoint.sol]: Would you mind adding a quick test for this where the other fixed point functions are tested?

Comment (chrismaree) [reply core/contracts/FixedPoint.sol]: sure. Will expand the tests when the next PR to expand the `fixedpoint` lib.
